### DISCUSSION TITLE
Searching for records in Budibase DB

### DISCUSSION THREAD

**Discussion by aderici** [Budibase/budibase#7140] [Help] [answered]
I'm in a sittuation where users need to search through a table of records. 

* I'm using budibase db. 
* The fields i'm searching through is a field like "title" or "description" where the search keywords can potentially be anywhere within the text.

I have read that this is not possible with budibase db, i want to double check that because i have seen a Clouseau installation within docker so shouldnt we be able to do this? 

If that's not going to be possible within the capabilities of budibase, i'd love to hear your ideas of a potential workaround.

Best regards,

**Comment by mike12345567** [comment]
Hi @aderici - sadly this is something that Clouseau + CouchDB isn't capable of, while it is possible to create full wildcard `*:*` lucene syntax and have Clouseau execute it, it has been disabled (`leadingWildcard` option) for use with CouchDB.

This is due to the fact that there isn't really a performant way to do this, other than performing string operations for every document in the database, if the database has millions of documents within it then it will have to search every single one of those documents, something that cannot be done in nominal time.

At the minute we have disabled tokenized search on our indexes because it doesn't fit with our needs exactly, however perhaps in the future we could have a think about dynamically creating tokenized indexes for a table if the user desires to search for strings split by say spaces e.g. `I want ice-cream` the user could search for `want` or `ice` and get the result due to the tokenized index.

I'm going to convert this to a discussion as I don't believe there is anything actionable for this right now and the community may have better suggestions for how to resolve the issue.

**Reply by aderici** [reply]
I dont think theres a problem with scanning individual records or an index. It all depends on the use-case and how crowded you expect the table to get. Mongodb has a regex match as well. Actually i wanted to switch to mongodb but then i can not use attachments feature.

Now i think i'll have to replicate the data to a third party service and use it to perform my search and populate my table. And then row click will actually lead to native budibase screen with budibase db data

**Reply by mike12345567** [reply]
Hi @aderici - yes I have to agree that under some circumstances it isn't a real issue, but sadly this is a decision that is out of the hands of Budibase, the Clouseau analysers do not expose the `setAllowLeadingWildcard` option which we would need to enable this and the decision has been made that it would be dangerous to allow this, so there isn't really a way for us to enable this without a bit of work to write out own analyser and build out our own version of the Clouseau analyzer that supports it.

If this work was picked up by the CouchDB/Clouseau maintainers, or someone else built us an analyser then we would happily support it, but at the minute its not something that we can easily support.

**Comment by aderici** [comment, marked as answer]
@mike12345567 I'm kind of new to couchdb but doesn't this simply mean we could do a regex comparison for a field? I'm self hosting so i'm ok to wait a few seconds or waste a few cpu cycles :

<img width="774" alt="image" src="https://user-images.githubusercontent.com/6830693/183462901-93681488-6a9a-4341-ad0e-82f2c87408d3.png">

because that's what i was doing in mongodb as well,

**Reply by mike12345567** [reply]
Hi @aderici - this is the `_find` API of CouchDB - it uses Mango queries and runs on view indexes but cannot be used in combination with the Clouseau Lucene index.

Currently we don't make use of mango queries anywhere as they require static indexes to run upon, or you run them on the `_all_docs` built in index and iterate over every document in the database. There is a world in which we could implement a very low performance regex option to fallback on (like a query for the built in tables) but its something that we've never done in Budibase and we would need to make the user aware of how poorly it will perform at scale. The reason we use Lucene instead of Mango is that we don't need to create an index for each of the users individual filters they create (if a Mango query is not indexed then it will need to be performed over every document, every time) whereas Lucene provides a dynamic search index that we can perform all sorts of filters on.

I believe there is something in the idea of providing an option to perform a regex based Mango query instead of using the standard filters, but this would need some significant thought, how to make the user aware it could timeout at scale, how it would fit into the UI, writing the Mango integration + all the sort, pagination etc options that we have on the Lucene index.

**Reply by aderici** [reply]
@mike12345567 I love budibase, i got %90 of the project figured out. But then there's this 10 percent we have to figure out to ensure project success because of an important requirement or end user expectation. The other day it was a clickable card, today it was search over a field or if i switch to another db type for search, being able to use attachments. 

I'm super happy with the interaction with team members and understanding the challenges as well as rationales. Just sometimes, we have to find a workaround so that our projects can continue and we can live to fight another day. 

Long story short, i have gone through couchdb documentation, and came up with this. I will implement this as a rest data source. It is good enough for my limited use case. For anyone else needing this you can try and see if it works for you : 

First we create an index because we want to get most recent created records first (thats my use-case) and i assume this is just a one-time task :

```
curl --location --request POST 'https://YOUR_BUDIBASE_URL_HERE/db/APP_ID_HERE/_index' \
-u "YOUR_COUCHDB_USER:YOUR_COUCHDB_PASSWORD"
--header 'Content-Type: application/json' \
--data-raw '{
  "index": { 
    "fields": [
      "Created At",
      "COLUMN_NAME_TO_SEARCH"
     ],
    "partial_filter_selector": { "_id": { "$regex": "^ro_ta_YOUR_TABLE_ID_HERE" } }
    },
    "ddoc":"MY_INDEX_NAME"
}'
```


Then we search our tables column and get the 50 most recent results (implement this as a rest data source)

```
curl --location --request POST 'https://YOUR_BUDIBASE_URL_HERE/db/APP_ID_HERE/_find' \
-u "YOUR_COUCHDB_USER:YOUR_COUCHDB_PASSWORD"
--header 'Content-Type: application/json' \
--data-raw '{
    "selector": {
        "_id": { "$regex": "TABLE_ID_HERE" },
        "COLUMN_NAME_TO_SEARCH": {"$regex": "SEARCH_STRING_HERE"}
    },
    "fields": ["_id", "_rev", "COLUMN_NAME_TO_SEARCH"],
    "sort": [{"Created At":"desc"}, {"COLUMN_NAME_TO_SEARCH": "desc"}],
    "limit": 50,
    "skip": 0,
    "use_index": "_design/MY_INDEX_NAME,
    "execution_stats": false
}'
```

Am i doing something terribly wrong here? It just takes 1 second for 3K records. I can live with this even if it takes 10. Seems to be good enough for my use case. I'd rather stay with this for now than going for an external component if there won't be any fire coming from my azure services.

**Reply by mike12345567** [reply]
Hi @aderici - yes this looks like a pretty good method to get going for now - nicely done working this out! Hopefully this will work for you long term, sadly 10 seconds is probably not the upper limit - these type of regex operations are incredibly slow, if your database ever reaches say 100K documents (not rows, as all documents are included in the search) then you could experience API request timeouts as generally these things get exponentially slower (caches filling, etc etc). However I will say, the use of the partial index for recently created documents is likely to provide relatively good speed assuming there is no major dump of documents - in theory this should keep you in the few thousand document area relatively consistently. Sadly this is just true of any full scan type search where every document has to be checked to return results, the only real efficient way to perform this is with a tokenized index - something we will think about and see how we could support.

As for getting stuck on the last 10% - this is something we realise can be problematic, there are a few things we're doing in this area to try and improve things/make it easier to extend Budibase:
1. Custom datasources - soon you'll be able to write your own custom datasources in JavaScript/TypeScript, in this case you could build out a Mango query datasource to use, or the community may have already had one built.
2. Custom components - I believe this will be more useful to you in the future, with this you are able to build totally custom components and import them into the component menus, so in the example of needing a clickable card, while this may be something we haven't built into the core product, it can be introduced as a plugin/shared with the community.

**Reply by aderici** [reply]
@mike12345567 just to confirm, i'm using a partial index that only indexes rows of my single table and i'm enforcing that index in my query. Should i still be worried about number of all documents in my db?

**Reply by mike12345567** [reply]
Hi @aderici - due to the use of the regex statement in the partial index, the index itself still has to run over every document in the database, however this is more of a slowdown in the write/update process, so you may only experience problems with large bulk updates. I don't think however there is anyway around this with Mango indexes, and as long as the table doesn't grow too large it should maintain a relative level of speed.

**Reply by aderici** [reply]
I see, still i don't think it as a big problem for now, it gives me some time and allows project to continue. I could avoid the regex as i know you include tableId field in all documents. 

This has been a very useful conversation and hopefully helps future visitors. Thank you @mike12345567 for all your help!

**Reply by mike12345567** [reply]
Hi @aderici glad I could help, always good to hear feedback on these things as well, see if we can come up with an efficient/easy way to do something within Budibase itself!
